### ISSUE TITLE
How would I go about using KernelExplainer for a sum product ensemble of 10 GAM models?

### ISSUE THREAD

**Issue by nickcorona** [shap/shap#925]
I work in insurance where we price frequency and severity of claims separate for each coverage in an insurance policy. Frequency is the number of times a customer claims, and severity is how severe a claim is.

Say there are 5 different coverage IDs: 1, 2, 3, 4, and 5. Any particular policy may or may not have any of the given coverages. And for each coverage, we train a frequency and severity model.

How we get to total loss (our target) for any observation is `severity_1 * frequency_1 + severity_2 * frequency_2 + severity_3 * frequency_3 + severity_4 * frequency_4 + severity_5 * frequency_5`.

I'm trying to figure out how I can create a prediction function or run a kernel explainer on each of the individual 10 models and somehow combine the SHAP values to get something sensible to explain why total losses for certain segments of the business are having large premium changes.

**Comment by slundberg** [comment]
A nice property of Shapley values is that if you linearly combine the outputs of models then you can linearly combine the SHAP values of those models in the same way to get the explanation of the final result. The problem with your case is that you have product (quadratic) terms so it is not quite so simple. You could just write a python function that does what you said and then explain that with KernelExplainer. But since you used "GAM" in the title I will assume you are using all neural network models. If that is the case then just define a new "joint" model in TF or PyTorch that computes the final loss and then explain that with GradientExplainer or DeepExplainer.

**Comment by nickcorona** [comment]
Sorry, by GAM I mean generalized additive model, it's not a neural network.

**Comment by slundberg** [comment]
@nickcorona ha! I must have been thinking GAN and writing GAM. If you are using GAM's check out the new shap.AdditiveExplainer it is fast and exact for first order GAM models. And since a linear combination of GAM models is still a GAM you can just use AdditiveExplainer for the whole ensemble :)

**Comment by nickcorona** [comment]
> @nickcorona ha! I must have been thinking GAN and writing GAM. If you are using GAM's check out the new shap.AdditiveExplainer it is fast and exact for first order GAM models. And since a linear combination of GAM models is still a GAM you can just use AdditiveExplainer for the whole ensemble :)

Sorry for the late reply.  Thanks a lot, Scott! I'll try this out and let you know how it goes.

Is there any documentation on the AdditiveExplainer? I don't see it on https://shap.readthedocs.io/.

**Comment by bhamel1** [comment]
@nickcorona I am facing a similar problem. Two models (both XGBoost) where one predicts the probability of an event and the other predicts the magnitude of the event. 

My function is 
expected value = (probability_of_event_1* magnitude_1) + (probability_of_event_2 * magnitude_2)

Do you have any code/example you can share for how you used the Additive explainer?

@slundberg Any documentation on how to use Shap with product sums? Or stacking several models together?
